src/IO/S3/Client.cpp

```diff
@@ -178,70 +178,70 @@ Model::HeadObjectOutcome Client::HeadObject(const HeadObjectRequest & request) c
 
     request.overrideURI(std::move(*bucket_uri));
 
-    return Aws::S3::S3Client::HeadObject(request);
+    return HeadObject(static_cast<const Model::HeadObjectRequest&>(request));
 }
 
 Model::ListObjectsV2Outcome Client::ListObjectsV2(const ListObjectsV2Request & request) const
 {
-    return doRequest(request, [this](const Model::ListObjectsV2Request & req) { return Aws::S3::S3Client::ListObjectsV2(req); });
+    return doRequest(request, [this](const Model::ListObjectsV2Request & req) { return ListObjectsV2(req); });
```

Comment:
`doRequest` calls a passed function multiple times. It seems you're making this `Client::ListObjectsV2` recursive here, it doesn't look correct.

Reply:
No. There is no recursion.
The functions
- Client::ListObjectsV2(const ListObjectsV2Request&) and 
- Aws::S3::S3Client::ListObjectsV2(const Model::ListObjectsV2Request &)
are different functions. They have different argument type. They are not overwrite each other. Aws::S3::S3Client::ListObjectsV2 is virtual, meanwhile Client::ListObjectsV2 is not.

Here I do a possibility to call ListObjectsV2(const Model::ListObjectsV2Request) by virtual call. This is needed to have an ability to override the virtual calls Aws::S3::S3Client::* for mocking client.

Reply:
I understand they're different functions, but it looks like `S3::Client::ListObjectsV2()` used to call `Aws::S3::S3Client::ListObjectsV2()`, and now with your changes `S3::Client::ListObjectsV2()` calls `S3::Client::ListObjectsV2()`.

Reply:
@antonio2368

Reply:
yeah it isn't very clear at first, but to me, the functions are simple enough and maybe add the comment above the first implementation (`ListObjectsV2`) something like `For each request, we wrap the request functions from Aws::S3::Client with doRequest`

Reply:
I don't feel good about this thing. Maybe we can at least add some comments?

Reply:
Made some comment here